sdk/python/lib/pulumi/runtime/known_types.py

```diff
@@ -12,294 +12,114 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """
-The known_types module contains state for keeping track of types that
-are known to be special in the Pulumi type system.
+The known_types module lazily loads classes defined in the parent module to
+allow for type checking.
 
 Python strictly disallows circular references between imported packages.
 Because the Pulumi top-level module depends on the `pulumi.runtime` submodule,
 it is not allowed for `pulumi.runtime` to reach back to the `pulumi` top-level
 to reference types that are defined there.
 
 In order to break this circular reference, and to be clear about what types
-the runtime knows about and treats specially, this module exports a number of
-"known type" decorators that can be applied to types in `pulumi` to indicate
-that they are specially treated.
+the runtime knows about and treats specially, we defer loading of the types from
+within the functions themselves.
 
-The implementation of this mechanism is that, for every known type, that type
-is stashed away in a global variable. Whenever the runtime wants to do a type
-test using that type (or instantiate an instance of this type), it uses the
-functions defined in this module to do so.
+This implementation allows for overriding types used as stubs for testing
+(see test/test_next_serialize.py)
 """
 from typing import Any, Optional
 
 _custom_resource_type: Optional[type] = None
```

Comment:
Any reason why we need to keep these global variables around?

I know `_custom_resource_type` is referenced by one of the tests:
 https://github.com/pulumi/pulumi/blob/dd104a00a76fec807fb5b50e11ceca37508757e1/sdk/python/lib/test/test_next_serialize.py#L50-L54

Reply:
@justinvp yeah that's the only reason that I kept them around (see the comment on line 27-28). I couldn't get one of the tests to work without it (test_custom_resource failed with a ValueError because it doesn't expect `FakeCustomResource`).
Is that test particularly useful? Is it worth keeping only the global for `CustomResource` and not the others?

Reply:
> Is it worth keeping only the global for `CustomResource` and not the others?

I don't think there's any reason to keep the others, so let's get rid of them. And then as a follow-up, see if we can rework the test to get rid of the `CustomResource` one.